### DISCUSSION TITLE
Can we do React-style iteration over `children`?

### DISCUSSION THREAD

**Discussion by danielkov** [DioxusLabs/dioxus#1818] [Q&A] [answered]
## Problem

In React, we can perform magic tricks with `React.Children`, e.g.: create a component that can take `asChild` prop that changes the component in that it "unwraps" the DOM tree and applies the props it would normally pass to a wrapper element to the first child instead, like so:

```tsx
const Example = () => {
  return (
    <Trigger asChild>
      <button className="myButton">
        Click me
      </button>
    </Trigger>
  );
}
```

would produce the following DOM tree:

```html
<button {...props from Trigger} class="myButton">
  Click me
</button>
```

by doing something like this in `Trigger`:

```tsx
const Trigger = ({
  children,
  asChild,
}: {
  children: React.ReactNode;
  asChild?: boolean;
}) => {
  const [show, setShow] = React.useState(false);

  if (asChild) {
    return React.Children.only(
      React.cloneElement(children as React.ReactElement, {
        "data-hidden": !show,
        onClick: () => setShow(true),
      })
    );
  }

  return <button onClick={() => setShow(true)}>{children}</button>;
};
```

## Question

Does Dioxus provide a similar API for utilities like `React.Children` and `React.cloneElement`? Would the following be possible?

```rs
fn Example(cx: Scope) -> Element {
    render! {
        Trigger {
            as_child: true,
            class: "myButton",
            button {
                "Click me!"
            }
        }
    }
}
```

Based on my digging around it seems like this could be possible, but I didn't want to dive too deep before I made sure there's not an official way of achieving something similar.

**Comment by ealmloff** [comment, marked as answer]
There has been some discussion about a VNode iterator [in this issue](https://github.com/DioxusLabs/dioxus/issues/1177#issuecomment-1629045604), but would be more complicated than react because of the [structure of VNodes](https://dioxuslabs.com/learn/0.4/contributing/walkthrough_readme#the-rsx-macro) that enables dioxus' improved performance.

You can [spread props](https://github.com/DioxusLabs/dioxus/blob/4f077ec664bf7a847e3c5e520b1c10fd506b3886/examples/spread.rs#L32) pretty easily in the git version of dioxus. I think the API you posted is possible, but it would be much easier to support an API like this:
```rust
fn Example(cx: Scope) -> Element {
    render! {
        button {
            class: "myButton",
            ..create_trigger(),
            "Click me!"
        }
    }
}
```
